src/firebolt_db/firebolt_async_dialect.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+from types import ModuleType
+from typing import Any, Iterator, List, Optional, Tuple
+
+import firebolt.async_db as async_dbapi
+from firebolt.async_db import Connection
+from sqlalchemy.engine import AdaptedConnection  # type: ignore[attr-defined]
+from sqlalchemy.util.concurrency import await_only
+
+from firebolt_db.firebolt_dialect import FireboltDialect
+
+
+class AsyncCursorWrapper:
+    __slots__ = (
+        "_adapt_connection",
+        "_connection",
+        "description",
+        "await_",
+        "_cursor",
+        "_rows",
+        "arraysize",
+        "rowcount",
+    )
+
+    server_side = False
+
+    def __init__(self, adapt_connection: AsyncConnectionWrapper):
+        self._adapt_connection = adapt_connection
+        self._connection = adapt_connection._connection
+        self.await_ = adapt_connection.await_
+        self.arraysize = 1
+        self.rowcount = -1
+        self.description = None
+        self._rows: List[Tuple] = []
+
+    def close(self) -> None:
+        self._rows[:] = []
+
+    def execute(self, operation: str, parameters: Optional[Tuple] = None) -> None:
+        _cursor = self._connection.cursor()
+        self.await_(_cursor.execute(operation, parameters))
+        if _cursor.description:
+            self.description = _cursor.description
+            self.rowcount = -1
+            self._rows = self.await_(_cursor.fetchall())
+        else:
+            self.description = None
+            self.rowcount = _cursor.rowcount
+
+        _cursor.close()
+
+    def executemany(self, operation: str, seq_of_parameters: List[Tuple]) -> None:
+        raise NotImplementedError("executemany is not supported yet")
+
+    def __iter__(self) -> Iterator[Tuple]:
+        while self._rows:
+            yield self._rows.pop(0)
+
+    def fetchone(self) -> Optional[Tuple]:
+        if self._rows:
+            return self._rows.pop(0)
+        else:
+            return None
+
+    def fetchmany(self, size: int = None) -> List[Tuple]:
+        if size is None:
+            size = self.arraysize
+
+        retval = self._rows[0:size]
+        self._rows[:] = self._rows[size:]
+        return retval
+
+    def fetchall(self) -> List[Tuple]:
+        retval = self._rows[:]
+        self._rows[:] = []
+        return retval
+
+
+class AsyncConnectionWrapper(AdaptedConnection):
+    await_ = staticmethod(await_only)
+    __slots__ = ("dbapi", "_connection")
+
+    def __init__(self, dbapi: AsyncAPIWrapper, connection: Connection):
+        self.dbapi = dbapi
+        self._connection = connection
+
+    def cursor(self) -> AsyncCursorWrapper:
+        return AsyncCursorWrapper(self)
+
+    def rollback(self) -> None:
```

Comment:
should we raise `NotImplementedError` here? Does sqlalchemy doc specifies this behaviour? If no, I'm fine with `pass`

Reply:
Asyncio docs are very vague, but this one suggests if there's no transaction it should have no effect: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncConnection.rollback